system/jlib/jevent.cpp

```diff
@@ -0,0 +1,560 @@
+//
+//   uncompressedSize   4 bytes
+//   compressedSize     4 bytes
+//   compressedData     compressedSize bytes
+//
+// The compression should be implemented by the streamed writing code.
+//
+// Should there be a terminator at the end of the file?  If will potentially cause problems if the file is compressed...
+
+EventRecorder::EventRecorder() : buffer(OutputBufferSize)
+{
+}
+
+void EventRecorder::checkAttrValue(EventAttr attr, size_t size)
+{
+    unsigned expectedSize = attrInformation[attr].size;
+    assertex(expectedSize == 0 || expectedSize == size);
+}
+
+void EventRecorder::startRecording(const char * optionsText, const char * filename)
+{
+    assertex(filename);
+    CriticalBlock block(cs);
+    if (isActive())
+        return;
+
+    auto processOption = [this](const char * option, const char * valueText)
+    {
+        bool valueBool = strToBool(valueText);
+
+        if (strieq(option, "traceid"))
+            options = (options & ~ERFtraceid) | (valueBool ? ERFtraceid : ERFnone);
+        else if (strieq(option, "threadid"))
+            options = (options & ~ERFthreadid) | (valueBool ? ERFthreadid : ERFnone);
+        else if (strieq(option, "stack"))
+            options = (options & ~ERFstacktrace) | (valueBool ? ERFstacktrace : ERFnone);
+    };
+
+    options = ERFthreadid; // set defaults
+    processOptionString(optionsText, processOption);
+    sizeMessageHeaderFooter = sizeof(EventType) + sizeof(__uint64) + sizeof(EventAttr); // event type, timestamp and end of attributes marker
+    if (options & ERFthreadid)
+        sizeMessageHeaderFooter += sizeof(__uint64);
+    if (options & ERFtraceid)
+        sizeMessageHeaderFooter += 32;
+
+    Owned<IFile> outputFile = createIFile(filename);
+    output.setown(outputFile->open(IFOcreate));
+
+    startCycles = get_cycles_now();
+    __uint64 startTimestamp = getTimeStampNowValue()*1000;
+
+    //Revisit: If the file is being compressed, then these fields should be output uncompressed at the head
+    offset_type pos = 0;
+    write(pos, currentVersion);
+    write(pos, options);
+    write(pos, startTimestamp);
+    nextOffset = pos;
+    nextWriteOffset = 0;
+    for (unsigned i=0; i < numBlocks; i++)
+        counts[i] = 0;
+
+    recordingEvents.store(true, std::memory_order_release);
+}
+
+void EventRecorder::stopRecording()
+{
+    if (!isActive())
+        return;
+
+    //MORE: Protect against startRecording() being called concurrently, by introducing another boolean to
+    //indicate if it is active, which is only cleared once this function completes.
+    recordingEvents.store(false);
+
+    //Need to wait until all writes have finished
+    for (unsigned iter=0; iter < 1000; iter++)
+    {
+        {
+            CriticalBlock block(cs);
+            if (getBlockFromOffset(nextOffset) == getBlockFromOffset(nextWriteOffset))
+                break;
+        }
+        MilliSleep(10);
+    }
```

Comment:
crazy unlikely no doubt, but if test in loop isn't met and has reached max iterations, should issue a warning at least?

Reply:
That is what the test on line 294 is meant to do - but the test was the wrong way around.  I will fix.